crates/pathfinder/src/sequencer.rs

```diff
@@ -1087,4 +1162,99 @@ mod tests {
     async fn eth_contract_addresses() {
         retry_on_rate_limiting!(client().eth_contract_addresses().await).unwrap();
     }
+
+    mod retry {
+        use super::{SequencerError, StarknetErrorCode};
+        use assert_matches::assert_matches;
+        use http::StatusCode;
+        use pretty_assertions::assert_eq;
+
+        async fn run_retry(
+            statuses: Vec<(StatusCode, &'static str)>,
+        ) -> Result<String, SequencerError> {
+            use http::response::Builder;
+            use std::{
+                cell::RefCell,
+                sync::{Arc, Mutex},
+            };
+            use warp::Filter;
+
+            let statuses = Arc::new(Mutex::new(RefCell::new(statuses)));
+            let any = warp::any().map(move || {
+                let s = statuses.clone();
+                let s = s.lock().unwrap();
+                let s = s.borrow_mut().pop().unwrap();
+                Builder::new().status(s.0).body(s.1)
+            });
+
+            let (addr, run_srv) = warp::serve(any).bind_ephemeral(([127, 0, 0, 1], 0));
+            let _jh = tokio::spawn(run_srv);
+
+            // super::retry is the UUT here
+            let result = super::retry(|| async {
+                let mut url = reqwest::Url::parse("http://localhost/").unwrap();
+                url.set_port(Some(addr.port())).unwrap();
+                let resp = reqwest::get(url).await.unwrap();
+                super::parse::<String>(resp).await
+            })
+            .await;
+            result
+        }
+
+        #[tokio::test]
+        async fn stop_on_ok() {
+            let ends_with_ok = vec![
+                (StatusCode::OK, r#""Finally!""#),
+                (StatusCode::REQUEST_TIMEOUT, ""),
+                (StatusCode::GATEWAY_TIMEOUT, ""),
+                (StatusCode::SERVICE_UNAVAILABLE, ""),
+                (StatusCode::BAD_GATEWAY, ""),
+                (StatusCode::TOO_MANY_REQUESTS, ""),
+            ];
+
+            let result = run_retry(ends_with_ok).await.unwrap();
+            assert_eq!(result, "Finally!");
+        }
+
+        #[tokio::test]
+        async fn stop_on_fatal() {
+            let ends_with_ok = vec![
+                (
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                    r#"{"code":"StarknetErrorCode.BLOCK_NOT_FOUND","message":""}"#,
+                ),
+                (StatusCode::REQUEST_TIMEOUT, ""),
+                (StatusCode::GATEWAY_TIMEOUT, ""),
+                (StatusCode::SERVICE_UNAVAILABLE, ""),
+                (StatusCode::BAD_GATEWAY, ""),
+                (StatusCode::TOO_MANY_REQUESTS, ""),
+            ];
```

Comment:
Same as for `OK` -- should test with the exit condition somewhere in the middle.